File: 0222-count-complete-tree-nodes/0222-count-complete-tree-nodes.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
   
    def countNodes(self, root: Optional[TreeNode]) -> int:
        def rightCount(node):
            return 0 if not node else 1 + rightCount(node.right)

        def leftCount(node):
            return 0 if not node else 1 + leftCount(node.left)

        def count(root):
            if not root:
                return 0

            left, right = leftCount(root), rightCount(root)
            if left == right:
                return 2**(left)-1

            return 1 + count(root.left) + count(root.right)
        return count(root)

File: 0222-count-complete-tree-nodes/0222-count-complete-tree-nodes.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int countNodes(TreeNode* root) {
        if(root == NULL){
            return 0;
        }
        
        int leftHeight=0, rightHeight=0;
        TreeNode* curr = root;
        
        while(curr!=NULL){
            leftHeight++;
            curr = curr->left;
        }
        
        curr = root;
        while(curr!=NULL){
            rightHeight++;
            curr = curr->right;
        }
        
        if(leftHeight==rightHeight){
            return pow(2, leftHeight)-1;
        }
        
        return 1 + countNodes(root->left) + countNodes(root->right);
    }
};